### PR TITLE
Do not polyfill node built-in modules on edge functions

### PR DESCRIPTION
**Pull request by nkzawa**
As the title. This is intended to be applied on both middleware and edge functions.

## Bug

- [ ] Related issues linked using `fixes #number`
- [x] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| buildDuration | 15.2s | 15s | -154ms |
| buildDurationCached | 6s | 6s | -43ms |
| nodeModulesSize | 484 MB | 484 MB | ⚠️ +1.74 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.006 | 3.003 | 0 |
| / avg req/sec | 831.67 | 832.46 | +0.79 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.164 | 1.169 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 2148.35 | 2138.54 | ⚠️ -9.81 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.3 kB | 28.3 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72 kB | 72 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.04 kB | 3.04 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.73 kB | 5.73 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| _buildManifest.js gzip | 461 B | 461 B | ✓ |
| Overall change | 461 B | 461 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| buildDuration | 18.1s | 18s | -129ms |
| buildDurationCached | 6s | 6s | ⚠️ +41ms |
| nodeModulesSize | 484 MB | 484 MB | ⚠️ +1.74 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.994 | 2.995 | 0 |
| / avg req/sec | 835.14 | 834.69 | ⚠️ -0.45 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.175 | 1.173 | 0 |
| /error-in-render avg req/sec | 2127.14 | 2131.61 | +4.47 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.7 kB | 28.7 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.6 kB | 72.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.02 kB | 3.02 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.77 kB | 5.77 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.44 kB | 2.44 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 393 B | 393 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | nkzawa/next.js middleware-no-node-buildin-module | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: c296a680bb4c5387abf6926db308de0bde980475

File: packages/next/build/index.ts [hunk 1]
```diff
@@ -739,6 +739,13 @@ export default async function build(
         )
       }
 
+      const breakingChangeIndex = error.indexOf(
+        '\n\nBREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.'
```
Comment (timneutkens) [inline packages/next/build/index.ts:743]: @sokra is there a better way to exclude the compilation from showing these, an option maybe?

**Review by timneutkens — DISMISSED** [review]
@nkzawa this is only the starting point right? Next step being improving the error shown when you import these in the edge runtime? If so the change looks good and it can be landed.

**Comment by nkzawa** [comment]
@timneutkens yes, we're going to improve error messages as well

**Comment by nkzawa** [comment]
Seems this causes to break edge runtime. We need to fix that too.

**Comment by nkzawa** [comment]
Changed to apply the fix only to middleware for now, because of the above problem

**Comment by ijjk** [comment]
## Failing test suites 
Commit: cd081163db4ea2ad74e06effa52137dedbe64ddd

`yarn testheadless test/e2e/middleware-can-use-wasm-files/index.test.ts` 
- middleware can use wasm files > uses the wasm file
- middleware can use wasm files > can be called twice
- middleware can use wasm files with the experimental modes on > uses the wasm file

<details>
<summary>Expand output</summary>

  ● middleware can use wasm files › uses the wasm file

    FetchError: invalid json response body at http://localhost:39015/ reason: Unexpected token < in JSON at position 0

      41 |   it('uses the wasm file', async () => {
      42 |     const response = await fetchViaHTTP(next.url, '/')
    > 43 |     expect(await response.json()).toEqual({
         |            ^
      44 |       input: 1,
      45 |       value: 2,
      46 |     })

      at ../node_modules/node-fetch/lib/index.js:273:32
      at Object.<anonymous> (e2e/middleware-can-use-wasm-files/index.test.ts:43:12)

  ● middleware can use wasm files › can be called twice

    FetchError: invalid json response body at http://localhost:39015/?input=2 reason: Unexpected token < in JSON at position 0

      49 |   it('can be called twice', async () => {
      50 |     const response = await fetchViaHTTP(next.url, '/', { input: 2 })
    > 51 |     expect(await response.json()).toEqual({
         |            ^
      52 |       input: 2,
      53 |       value: 3,
      54 |     })

      at ../node_modules/node-fetch/lib/index.js:273:32
      at Object.<anonymous> (e2e/middleware-can-use-wasm-files/index.test.ts:51:12)

  ● middleware can use wasm files with the experimental modes on › uses the wasm file

    FetchError: invalid json response body at http://localhost:36443/ reason: Unexpected token < in JSON at position 0

       96 |   it('uses the wasm file', async () => {
       97 |     const response = await fetchViaHTTP(next.url, '/')
    >  98 |     expect(await response.json()).toEqual({
          |            ^
       99 |       input: 1,
      100 |       value: 2,
      101 |     })

      at ../node_modules/node-fetch/lib/index.js:273:32
          at runMicrotasks (<anonymous>)
      at Object.<anonymous> (e2e/middleware-can-use-wasm-files/index.test.ts:98:12)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/build/index.ts (hunk 1 above)
Comment (nkzawa) [reply packages/next/build/index.ts:743]: looks there is no other way as far as I checked https://github.com/webpack/webpack/blob/main/lib/ModuleNotFoundError.js#L63

**Review by huozhi — DISMISSED** [review]

**Review by shuding — APPROVED** [review]

**Pull request merged**